core/src/main/java/io/grpc/internal/ClientCallImpl.java

```diff
@@ -326,7 +327,13 @@ public void onReady() {
       callExecutor.execute(new Runnable() {
         @Override
         public void run() {
-          observer.onReady();
+          try {
+            inOnReady = true;
```

Comment:
We're now re-entrant safe and we don't need to protect against being multi-threaded in this runnable as the executor is serializing